src/ocl/utilocl.cpp

```diff
@@ -565,8 +565,15 @@ float Col2Im3dGPU(const Handle& handle,
     }
     else
     {
+        std::size_t index_size = static_cast<size_t>(in_c) * out_d * out_h * out_w * wei_d * wei_w *
+                                 wei_h * sizeof(ConstData_t);
+
+        const bool use_64_bit_index = index_size > 0xffffffffULL;
```

Comment:
@zjing14 please comment. Thanks!

Reply:
@zjing14 WRT question 1. AFAICS `sizeof(ConstData_t)` it should be `GetTypeSize(type)`, which matches `sizeof(_FLOAT)` in the kernel, right? It should be either 4 (fp32) or 2 (fp16), I think.

Question 2 remains the same.

Reply:
> [Performance]
> 
> 1. Hmm, why `* sizeof(ConstData_t)`, which is `sizeof(void*) == 8` for HIP backend? In the shader,
> 
> ```
> ch_offset = im_ch * col_d * col_w * col_h * wei_d * wei_w * wei_h
> ```
> 
> where max value of im_ch is in_c, right?
> 
> Multiplying this to 8 looks like an overkill.

Yes, it should be  `sizeof(ConstData_t)`.

> 2. Another question is about `0xffffffffULL` (4 GiB - 1). In the kernel, computations are performed as `int`, which has max value of `0x7fffffffULL` (2 GiB - 1).

Yes, `0x7fffffffULL` (2 GiB - 1) makes sense or we change it to `unsigned int`.

Reply:
@zjing14 Thanks for answering!
> Yes, it should be sizeof(ConstData_t).

Excuse me, but I do not understand why. AFAIU the goal is to prevent overflows in the following computations in the shader:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/7ad167d2ecdc29ccebdf965055e1b1a8b9a835a8/src/kernels/MIOpenCol2Im3d.cl#L87-L91

and
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/7ad167d2ecdc29ccebdf965055e1b1a8b9a835a8/src/kernels/MIOpenCol2Im3d.cl#L111-L119
For computation of `ch_offset` we do not need the `* sizeof(ConstData_t)` multiplier at all.

Is this multiplier necessary to prevent overflow during computation of `col_off`?

Reply:
Yes. You are right, the goal of `int64_t` is to prevent index overflow. We need to make sure all indices `ch_offset `and `col_off ` do not overflow.

Reply:
We may change indices in the kernel to `unsigned int` for 32bit. It may improve performance for some cases where indices > 16bit but < 32bit.

Reply:
@zjing14 Thanks for feedback. And yes, I'll change to unsigned. But what about removal of `* sizeof(ConstData_t)` from the computation of index_size? Is it Ok to do that or I am missing something?